src/InclusiveScans.jl

```diff
@@ -1,12 +1,188 @@
+# - Each block processes 2 * BLOCK_SIZE elements (2 per thread).
+# - If `blockSums !== nothing`, writes the "block sum" to blockSums[blockIdx.x].
+# - Otherwise skips this step (when we scan blockSums itself).
+function _scanBlockKernel!(
+    g_odata::CuDeviceVector{Float32,1},
+    g_idata::CuDeviceVector{Float32,1},
+    blockSums::Union{CuDeviceVector{Float32,1},Nothing},
+    n::Int,
+)
+    # Dynamic shared memory on 2*BLOCK_SIZE float
+    temp = @cuDynamicSharedMem(Float32, blockDim().x * 2)
+
+    tx = threadIdx().x - 1
+    bx = blockIdx().x - 1
+
+    # Every block evaluate 2*BLOCK_SIZE => start = bx*(2*BLOCK_SIZE)
+    start = bx * (blockDim().x * 2)
+
+    function sm(i::Int)
+        return temp[i+1]
+    end
+    function set_sm(i::Int, v::Float32)
+        temp[i+1] = v
+    end
+
+    i1 = start + tx * 2
+    i2 = i1 + 1
+
+    if tx >= 0 && 2 * tx < 2 * blockDim().x
+        if i1 < n
+            set_sm(2 * tx, g_idata[i1+1])
+        else
+            set_sm(2 * tx, 0.0f0)
+        end
+        if i2 < n
+            set_sm(2 * tx + 1, g_idata[i2+1])
+        else
+            set_sm(2 * tx + 1, 0.0f0)
+        end
+    end
+    sync_threads()
+
+    # Up-sweep (reduce)
+    offset = 1
+    d = blockDim().x
+    while d > 0
+        sync_threads()
+        if tx < d
+            ai = offset * (2 * tx + 1) - 1
+            bi = offset * (2 * tx + 2) - 1
+            set_sm(bi, sm(bi) + sm(ai))
+        end
+        offset <<= 1
+        d >>= 1
+    end
+
+    sync_threads()
+
+    # Save sum of block -> blockSums[bx], 
+    # Set last to 0 (to get EXCLUSIVE)
+    if tx == 0
+        if blockSums !== nothing
+            blockSums[bx+1] = sm(2 * blockDim().x - 1)
+        end
+        set_sm(2 * blockDim().x - 1, 0.0f0)
+    end
+    sync_threads()
+
+    # Down-sweep
+    d = 1
+    while d < 2 * blockDim().x
+        offset >>= 1
+        sync_threads()
+        if tx < d
+            ai = offset * (2 * tx + 1) - 1
+            bi = offset * (2 * tx + 2) - 1
+            t = sm(ai)
+            set_sm(ai, sm(bi))
+            set_sm(bi, sm(bi) + t)
+        end
+        d <<= 1
+    end
+    sync_threads()
+
+    if tx >= 0 && 2 * tx < 2 * blockDim().x
+        if i1 < n
+            g_odata[i1+1] = sm(2 * tx)
+        end
+        if i2 < n
+            g_odata[i2+1] = sm(2 * tx + 1)
+        end
+    end
+
+    return nothing
+end
+
+# Add prefix sum from previous blocks (d_increments[blockIdx.x]) to all elements processing by this block
+function _addIncrementsKernel!(
+    g_odata::CuDeviceVector{Float32,1},
+    incr::CuDeviceVector{Float32,1},
+    n::Int,
+)
+    tx = threadIdx().x - 1
+    bx = blockIdx().x - 1
+
+    start = bx * (blockDim().x * 2)
+    i = start + tx * 2
+
+    if i < n
+        g_odata[i+1] += incr[bx+1]
+        if i + 1 < n
+            g_odata[i+2] += incr[bx+1]
+        end
+    end
+    return nothing
+end
+
+function _exclusiveToInclusiveKernel!(
```

Comment:
For the next PR, you should change that the code is directly doing an inclusive scan instead an exclusive scan and then shifting the data. Memory copies are expensive.